styx-common/src/main/java/com/spotify/styx/util/ShardedCounter.java

```diff
@@ -0,0 +1,364 @@
+          }
+          return null;
+        });
+      }
+    }
+
+    /**
+     * Gets the counter state. Creates it in Datastore if it doesn't exist yet.
+     */
+    private static CounterSnapshot fromDatastore(Datastore datastore, String counterId, Time time) {
+      return new CounterSnapshot(datastore, counterId, time);
+    }
+
+    private static Map<Integer, Long> fetchShards(Datastore datastore, String counterId) {
+      final EntityQuery queryShards = EntityQuery.newEntityQueryBuilder()
+          .setKind(KIND_COUNTER_SHARD)
+          .setFilter(PropertyFilter.eq(PROPERTY_COUNTER_ID, counterId))
+          .setOrderBy(OrderBy.asc("__key__"))
+          .setLimit(NUM_SHARDS)
+          .build();
+      final QueryResults<Entity> shards = datastore.run(queryShards);
+      final Map<Integer, Long> fetchedShards = new HashMap<>();
+      while (shards.hasNext()) {
+        Entity shard = shards.next();
+        fetchedShards
+            .put((int) shard.getLong(PROPERTY_SHARD_INDEX), shard.getLong(PROPERTY_SHARD_VALUE));
+      }
+      return fetchedShards;
+    }
+
+    private boolean isRecent() {
+      return updatedAt.plus(CACHE_EXPIRY_DURATION.toEpochMilli(), ChronoUnit.MILLIS)
+          .isAfter(time.get());
+    }
+
+    /**
+     * The shard's capacity is calculated as:
+     * 1/NUM_SHARDS part of the total count capacity
+     * plus an extra 1 unit of the remainder (whenever limit % NUM_SHARDS > 0) depending on the shardIndex
+     * The first (limit % NUM_SHARDS) shards get 1 unit extra, and the rest get 0.
+     *
+     * </p>ex. If limit=5 for a given counter and NUM_SHARDS=3,
+     * then the distribution of capacity between the 3 shards will be [2, 2, 1]
+     */
+    private long shardCapacity(int shardIndex) {
+      // The shard's capacity is calculated as:
+      // 1/NUM_SHARDS part of the total count capacity
+      // plus an extra 1 unit of the remainder (whenever limit % NUM_SHARDS > 0) depending on the shardIndex
+      return limit / NUM_SHARDS + (shardIndex < limit % NUM_SHARDS ? 1 : 0);
+    }
+
+    /**
+     * Returns shard index which _likely_ could be successfully updated by delta, according to our
+     * cached view of the state in Datastore.
+     */
+    private int pickShardWithSpareCapacity(long delta) throws InvalidShardStateException {
+      List<Integer> candidates = new ArrayList<>();
+
+      for (int i : shards.keySet()) {
+        if (shards.containsKey(i) && Range
+            .closed(0L, shardCapacity(i)).contains(shards.get(i) + delta)) {
+          candidates.add(i);
+        } else if (!shards.containsKey(i)) {
+          LOG.warn("Shard {} for counter {} is not present in local cache", i, counterId);
+        }
+      }
+
+      if (candidates.isEmpty()) {
+        if (shards.size() == 0) {
+          throw new InvalidShardStateException(
+              "Trying to operate with a potentially uninitialized counter. Cache needs to be updated first.");
```

Comment:
This is a permitted state in normal operation. We may have updated the cache but still gotten zero shards. Let's not throw... In the earlier review I was still in favor of `return new Random().nextInt(NUM_SHARDS);` because that leads to fewer code paths (harder to make mistakes when maintaining) and only one place to throw CounterCapacityException, in updateCounter.